### PR TITLE
Add torch.bincount op lowering in linalg_ext

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This patch adds code to lower the op using the
`generateScalarImplementation` method.

Op description is given at
https://pytorch.org/docs/stable/generated/torch.bincount.html.

### CODE REVIEW

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 1]
```diff
@@ -1131,6 +1191,92 @@ Operation *ReverseOp::getTiledImplementation(OpBuilder &builder,
...
+    auto weightsType = op.weights().getType().cast<ShapedType>();
+    ArrayRef<int64_t> weightsShapes = weightsType.getShape();
+    if (weightsShapes.size() != 1) {
+      return op.emitError("expected weights rank to be one");
+    }
+    if (inputShapes[0] != weightsShapes[0]) {
+      return op.emitOpError("incompatible bins and weights shapes");
+    }
+    auto outputType = op.output().getType().cast<ShapedType>();
+    // If weights are given as input, the output must be floating-point type.
```
Comment (harsh-nod) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1226]: Shouldn't the element type of output be the same as weights? So if weights are i32, output i32? And if weights are f32, output is f32?

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 2]
```diff
@@ -1131,6 +1191,92 @@ Operation *ReverseOp::getTiledImplementation(OpBuilder &builder,
...
+  Value output_ = output();
+  Type outputType = output_.getType().cast<ShapedType>().getElementType();
+  Value one = b.create<arith::ConstantOp>(loc, b.getIntegerAttr(outputType, 1));
+  Value bin = b.create<memref::LoadOp>(loc, input_, ivs.front());
+  bin = castIntToIndex(b, loc, bin);
+  Value count = b.create<memref::LoadOp>(loc, output_, bin);
+  if (getNumInputs() == 2) {
+    Value weights_ = weights();
+    Value weight = b.create<memref::LoadOp>(loc, weights_, ivs.front());
+    weight = convertScalarToDtype(b, loc, weight, outputType);
```
Comment (harsh-nod) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1269]: why do you need to do convertScalarToDtype? If we require weight and output to be the same type does this go away?

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 1 above)
Comment (vivekkhandelwal1) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1226]: I checked this in python. When weights are given as input, the output is always a floating-point, irrespective of the type of weight.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 2 above)
Comment (vivekkhandelwal1) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1269]: This comment is linked to your previous comment, and I have answered why we can't put a constraint for the output and weights to be of the same type.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 1 above)
Comment (harsh-nod) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1226]: Makes sense to stick with torch's conventions. Looks good!

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp (hunk 2 above)
Comment (harsh-nod) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1269]: Sounds good!

**Review by harsh-nod — APPROVED** [review]

**Pull request closed without merging**